pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,194 @@
+	DelayTime       = time.Minute * 5
+)
+
+type PodStore struct {
+	curPod       cache.Indexer
+	prevPod      cache.Indexer
+	podsToDelete workqueue.DelayingInterface
+	// Mapping pod.uuid to LocalClock
+	timestampMap map[string]*LocalClock
+	clock        clock.WithTicker
+	mutex        sync.Mutex
+}
+
+type LocalClock struct {
+	// The local clock when the Pod is added
+	CreationTimestamp time.Time
+	// The local clock when the Pod is actually deleted from kube-apiserver. Nil if it's not deleted yet.
+	DeletionTimestamp *time.Time
+}
+
+// NewPodStoreWithClock creates a Pod Store with a custom clock,
+// which is useful when writing robust unit tests.
+func NewPodStoreWithClock(podInformer cache.SharedIndexInformer, clock clock.WithTicker) *PodStore {
+	s := &PodStore{
+		curPod:       podInformer.GetIndexer(),
+		prevPod:      cache.NewIndexer(podKeyFunc, cache.Indexers{PodIPIndex: podIPIndexFunc}),
+		podsToDelete: workqueue.NewDelayingQueueWithCustomClock(clock, DeleteQueueName),
+		clock:        clock,
+		timestampMap: map[string]*LocalClock{},
+	}
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.onPodCreate,
+			DeleteFunc: s.onPodDelete,
+		})
+	s.curPod.AddIndexers(cache.Indexers{PodIPIndex: podIPIndexFunc})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return NewPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodCreate(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Error while processing Create Event")
+		return
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	s.timestampMap[string(pod.UID)] = &LocalClock{CreationTimestamp: timeNow}
+	klog.V(4).InfoS("Process Pod Create Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) onPodDelete(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		var err error
+		pod, err = s.checkDeletedPod(obj)
+		if err != nil {
+			klog.ErrorS(err, "Got error while processing Delete Event")
+			return
+		}
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	s.timestampMap[string(pod.UID)].DeletionTimestamp = &timeNow
+	klog.V(4).InfoS("Process Pod Delete Event", "Pod", klog.KObj(pod))
+	s.prevPod.Add(pod)
+	s.podsToDelete.AddAfter(pod, DelayTime)
+}
+
+func (s *PodStore) checkDeletedPod(obj interface{}) (*corev1.Pod, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("received unexpected object: %v", obj)
+	}
+	pod, ok := deletedState.Obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Pod: %v", deletedState.Obj)
+	}
+	return pod, nil
+}
+
+func (s *PodStore) GetPodByIPAndTime(ip string, startTime time.Time) (*corev1.Pod, bool) {
+	// Query current Pod first
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	curPods, _ := s.curPod.ByIndex(PodIPIndex, ip)
+	if len(curPods) != 0 {
+		pod := curPods[0].(*corev1.Pod)
```

Comment:
why don't you iterate over all available Pods? One interest of this new Pod store is that if the IP address was reused, you can find the correct Pod. So you should look at all the Pods in `curPods` until you find a matching timestamp. Am I missing something?

Reply:
Is it possible that we have two pods in the indexer with same ip? I thought a Pod must be deleted before its IP is reused on another Pod?

Reply:
oh right, I thought this was the other indexer, for previous Pods
I am not sure it is possible. But then again, I would probably still iterate anyway and maybe use a closure to unify the code for both indexers.

Reply:
I've fixed it and also created our own indexer for `curPods` instead of using default indexer in PodInformer to keep consistency and avoid the issue below.
> 	// Might not be able to find previous Pod in prePods if prevPod.ByIndex is called immediately after
	// Delete Event happened. The clientState will delete the Pod in the informer before the DeleteFunc is called.